src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletActor.scala

```diff
@@ -418,68 +418,76 @@ class ErgoWalletActor(settings: ErgoSettings, boxSelector: BoxSelector)
-              require(prover.pubKeys.nonEmpty, "No public keys in the prover to extract change address from")
-              require(requests.count(_.isInstanceOf[AssetIssueRequest]) <= 1, "Too many asset issue requests")
-              require(payTo.forall(c => c.value >= BoxUtils.minimalErgoAmountSimulated(c, parameters)), "Minimal ERG value not met")
-              require(payTo.forall(_.additionalTokens.forall(_._2 >= 0)), "Negative asset value")
-
-              val assetIssueBox = payTo
-                .zip(requests)
-                .filter(_._2.isInstanceOf[AssetIssueRequest])
-                .map(_._1)
-                .headOption
-
-              val targetBalance = payTo
-                .map(_.value)
-                .sum
-
-              val targetAssets = payTo
-                .filterNot(bx => assetIssueBox.contains(bx))
-                .foldLeft(Map.empty[ModifierId, Long]) { case (acc, bx) =>
-                  // TODO optimize: avoid toArray and use mapFirst
-                  val boxTokens = bx.additionalTokens.toArray.map(t => bytesToId(t._1) -> t._2).toMap
-                  AssetUtils.mergeAssets(boxTokens, acc)
-                }
-
-              val (inputBoxes, filter) = if (inputs.nonEmpty) {
-                //inputs are provided externally, no need for filtering
-                (boxesToFakeTracked(inputs), noFilter)
-              } else {
-                //inputs are to be selected by the wallet
-                (registry.readCertainUnspentBoxes.toIterator, onChainFilter)
-              }
+        val inputs = inputsRaw
+          .toList
+          .map(in => Base16.decode(in).flatMap(ErgoBoxSerializer.parseBytesTry))
+          .map(_.get)
+
+        val dataInputs = dataInputsRaw
+          .toIndexedSeq
+          .map(in => Base16.decode(in).flatMap(ErgoBoxSerializer.parseBytesTry))
+          .map(_.get)
+
+        requestsToBoxCandidates(requests).flatMap { payTo =>
+          require(prover.pubKeys.nonEmpty, "No public keys in the prover to extract change address from")
+          require(requests.count(_.isInstanceOf[AssetIssueRequest]) <= 1, "Too many asset issue requests")
+          require(payTo.forall(c => c.value >= BoxUtils.minimalErgoAmountSimulated(c, parameters)), "Minimal ERG value not met")
+          require(payTo.forall(_.additionalTokens.forall(_._2 >= 0)), "Negative asset value")
+
+          val assetIssueBox = payTo
+            .zip(requests)
+            .filter(_._2.isInstanceOf[AssetIssueRequest])
+            .map(_._1)
+            .headOption
+
+          val targetBalance = payTo
+            .map(_.value)
+            .sum
+
+          val targetAssets = payTo
+            .filterNot(bx => assetIssueBox.contains(bx))
+            .foldLeft(Map.empty[ModifierId, Long]) { case (acc, bx) =>
+              // TODO optimize: avoid toArray and use mapFirst
+              val boxTokens = bx.additionalTokens.toArray.map(t => bytesToId(t._1) -> t._2).toMap
```

Comment:
There is the same code in wallet: 
https://github.com/ergoplatform/ergo/blob/f471e448b131d791cc4061f3168e97eb8a1517f0/ergo-wallet/src/main/scala/org/ergoplatform/wallet/transactions/TransactionBuilder.scala#L30-L31

Maybe it's time to extract and re-use it. Also `AssetUtils` can be moved to wallet and used in place of 
https://github.com/ergoplatform/ergo/blob/f471e448b131d791cc4061f3168e97eb8a1517f0/ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/BoxSelector.scala#L55-L78

Reply:
@greenhat done!